File: tools/shp_exporter.py
#! /usr/bin/python

import sys
import os
import argparse
import ogr
import osr

class SHPExporter:

    FAIMS_PROJ_WKT = 'PROJCS["WGS 84 / Pseudo-Mercator",GEOGCS["Popular Visualisation CRS",DATUM["Popular_Visualisation_Datum",SPHEROID["Popular Visualisation Sphere",6378137,0,AUTHORITY["EPSG","7059"]],TOWGS84[0,0,0,0,0,0,0],AUTHORITY["EPSG","6055"]],PRIMEM["Greenwich",0,AUTHORITY["EPSG","8901"]],UNIT["degree",0.01745329251994328,AUTHORITY["EPSG","9122"]],AUTHORITY["EPSG","4055"]],UNIT["metre",1,AUTHORITY["EPSG","9001"]],PROJECTION["Mercator_1SP"],PARAMETER["central_meridian",0],PARAMETER["scale_factor",1],PARAMETER["false_easting",0],PARAMETER["false_northing",0],AUTHORITY["EPSG","3785"],AXIS["X",EAST],AXIS["Y",NORTH]]'

    @staticmethod
    def export_dir(dirname, verbose):
        if os.path.isdir(dirname) == False:
            print "Error: " + dirname + " does not exist"
            return

        if dirname.endswith("/") == False:
            dirname = dirname + "/"

        for root, dirs, files in os.walk(dirname): # Walk directory tree
            for f in files:
                if f.endswith(".shp"):

                    SHPExporter.export_file(dirname, f, verbose)

    @staticmethod
    def export_file(dirname, filename, verbose):

        filepath = dirname + filename
        filepathNoExt = os.path.splitext(filepath)[0]

        if os.path.isfile(filepath) == False:
            print "Error: " + filename + " does not exist"
            return

        # get shape projection
        f = open(filepathNoExt + ".prj")
        proj_wkt = f.read()
        f.close()

        # create transform
        sRef = osr.SpatialReference()
        dRef = osr.SpatialReference()

        sRef.ImportFromWkt(proj_wkt)
        dRef.ImportFromWkt(SHPExporter.FAIMS_PROJ_WKT)

        #print sRef.ExportToWkt()
        #print dRef.ExportToWkt()

        transform = osr.CoordinateTransformation(sRef, dRef)

        # export file
        ds = ogr.Open(filepath)
        if ds is None:
            print "Error: Processing file " + filename + " failed"
            sys.exit( 1 )

        f = open(filepathNoExt + ".wkb", 'wb')

        count = 0;
        for i in range(ds.GetLayerCount()):
            #print "Layer: " + str(i)

            lyr = ds.GetLayerByIndex(i)
            lyr.ResetReading()

            for feat in lyr:
                geom = feat.GetGeometryRef()
                if geom == None:
                    continue

                count = count + 1

                geom.Transform(transform)

                #print geom.ExportToWkt()

                wkb = geom.ExportToWkb()
                f.write(wkb)

            if verbose:
                print "Exported file " + filename + " with " + str(count) + " geometry objects"

        f.close()
        ds = None

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Export .shp file[s] to .wkb file[s]")

    parser.add_argument("filename", metavar="file",
                      help="export .shp file to .wkb file")
    parser.add_argument("-d", "--dir",
                      action="store_true", dest="dirname", default=False,
                      help="export all files in directory")
    parser.add_argument("-q", "--quiet",
                      action="store_false", dest="verbose", default=True,
                      help="don't print status messages to stdout")

    namespace, extra = parser.parse_known_args()

    if namespace.dirname:
        SHPExporter.export_dir(namespace.filename, namespace.verbose)
    else:
        SHPExporter.export_file("./", namespace.filename, namespace.verbose)

File: tools/shp_exporter.rb
require "thor"

PROJ = "+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +no_defs"

class SHPExporter

  def self.export_dir(dirname)
    if !Dir.exists? dirname
      puts "ERROR: directory #{dirname} does not exist"
      return
    end
    dirname = dirname + '/' unless dirname[-1] == '/'
    `rm -f #{dirname}*_out.*`
    Dir.glob("#{dirname}*.shp") do |shp_file|
      export_file(dirname, shp_file)
    end
  end

  def self.export_file(dirname, filename)
    if !File.exists? filename or filename == "."
      puts "ERROR: file #{filename} does not exist"
      return
    end
    puts "Exporting #{filename}"
    `rm -f #{File.basename(filename, ".*")}_out.*`
    `ogr2ogr -t_srs "#{PROJ}" #{dirname + File.basename(filename, ".*") + "_out.shp"} #{filename}`
  end

end

class SHPExporterThor < Thor
  desc "export", "export and transform shape files"
  method_option :dir, :aliases => "-d",
                :desc => "export directory"
  method_option :quite, :aliases => "-q",
                :desc => "silent mode"
  def export(filename)
    if options[:dir]
      SHPExporter.export_dir(filename)
    else
      SHPExporter.export_file("./", filename)
    end
  end

end

if __FILE__ == $0
  SHPExporterThor.start(ARGV)
end
